### ISSUE TITLE
Image rebuild forced by different tags of a generic dev image

### ISSUE THREAD

**Issue by ashirokikh** [microsoft/navcontainerhelper#2664]
**Describe the issue**

The New-NavImage script rebuilds an image when ArtifactUrl and ImageName specified. A generic dev image is specified as the base image. The condition that forces image rebuild is as follows:

```powershell
      $labels = Get-BcContainerImageLabels -imageName $baseImage -registryCredential $registryCredential

      <...>

      elseif ($inspect.Config.Labels.tag -ne $labels.tag) {
          Write-Host "Image $imageName has generic Tag $($inspect.Config.Labels.tag), should be $($labels.tag)"
          $forceRebuild = $true
      }
```

However, an image is build with a tag that is received with the following line.

```powershell
    $genericTag = [Version](Get-BcContainerGenericTag -containerOrImageName $baseImage)
```

In this example, `$genericTag` has the value `1.0.2.4` but the tag from the image `$labels.tag` is `1.0.2.13`. 

This can be checked without container creation as follows:

```powershell
$baseImage = "$(Get-BestGenericImageName)-dev"
$registryCredential = $null

$genericTag = [Version](Get-BcContainerGenericTag -containerOrImageName $baseImage)
$labels = Get-BcContainerImageLabels -imageName $baseImage -registryCredential $registryCredential

Write-Host "Generic tag: $genericTag"
Write-Host "Image labels tag: $($labels.tag)"
```

The result is as follows:
```
BcContainerHelper version 3.0.12
BcContainerHelper emits usage statistics telemetry to Microsoft
INFO: Windows 10 21H1/21H2 images are not yet available, using 2004 as these are found to work better than 20H2 on 21H1/21H2
Generic tag: 1.0.2.4
Image labels tag: 1.0.2.13
```

This appears to be **an issue only when the `-useGenericImage` parameter is used**. It could be an issue only with the "$(Get-BestGenericImageName)-dev" image. Running the container creation script without the parameter line does not result in forced image build.

**Scripts used to create container and cause the issue**

```
$artifactUrl = "https://bcartifacts.azureedge.net/sandbox/20.0.37253.44689/au"
$containerName = "bc20auTest"
$authenticationType = 'NavUserPassword'

$credential = Get-Credential -Message "Please enter username and password for your container" -UserName "admin"

Measure-command {
    New-BcContainer -accept_eula `
                    -containerName $containerName `
                    -artifactUrl $artifactUrl `
                    -auth $authenticationType `
                    -credential $credential `
                    -updateHosts `
                    -useGenericImage "$(Get-BestGenericImageName)-dev" `
                    -imageName bc20auTest-image
}
```

**Full output of scripts**
```
INFO: Windows 10 21H1/21H2 images are not yet available, using 2004 as these are found to work better than 20H2 on 21H1/21H2
BcContainerHelper is version 3.0.12
BcContainerHelper is not running as administrator
UsePsSession is False
Host is Microsoft Windows 10 Pro - 21H2
Docker Client Version is 20.10.17
Docker Server Version is 20.10.17
Removing Session bc20auTest
Removing container bc20auTest
Removing entries from hosts
Removing bc20auTest from container hosts file
Removing bc20auTest-* from container hosts file
Removing Desktop shortcuts
Removing C:\ProgramData\BcContainerHelper\Extensions\bc20auTest
Fetching all docker images
Fetching all docker volumes
ArtifactUrl and ImageName specified
Image bc20autest-image:sandbox-20.0.37253.44689-au-mt already exists
Image bc20autest-image:sandbox-20.0.37253.44689-au-mt has generic Tag 1.0.2.4, should be 1.0.2.13
Building multitenant image bc20autest-image:sandbox-20.0.37253.44689-au-mt based on mcr.microsoft.com/businesscentral:10.0.19041.1415-dev with https://bcartifacts.azureedge.net/sandbox/20.0.37253.44689/au
Generic Tag: 1.0.2.4
Container OS Version: 10.0.19041.1415 (2004)
Host OS Version: 10.0.19044.1889 (21H2)
WARNING: Host OS is 21H2 and Container OS is 2004, defaulting to process isolation. If you experience problems, add -isolation hyperv.
Using process isolation
Files in c:\bcartifacts.cache\5b4ugicm.olw\my:
Copying Platform Artifacts
c:\bcartifacts.cache\sandbox\20.0.37253.44689\platform
Copying Database
Copying Licensefile
Copying ConfigurationPackages
C:\bcartifacts.cache\sandbox\20.0.37253.44689\au\ConfigurationPackages
Copying Extensions
C:\bcartifacts.cache\sandbox\20.0.37253.44689\au\Extensions
Copying Applications.AU
C:\bcartifacts.cache\sandbox\20.0.37253.44689\au\Applications.AU
Untagged: bc20autest-image:sandbox-20.0.37253.44689-au-mt
Deleted: sha256:e1cd0487643957ad167657a58ec80b5ccc225d2ef47a7852503fb1a7478b2ea3
Deleted: sha256:d3c2276eeb1406ec1908925de591d45d92d98efa37f02198c7489ea09efd3ba0
...
```

**Comment by freddydk** [comment]
could you try to use -alwaysPull once - to see whether that solves the issue.
Running this:
```
$labels = Get-BcContainerImageLabels -imageName 'mcr.microsoft.com/businesscentral:10.0.19041.1415-dev'
$labels.tag
1.0.2.13
```
Causes the rebuild to be flagged - but maybe it then uses the cached image to build the new image (which then again will be too old)

**Comment by freddydk** [comment]
The real problem is in this code:

```
    $hostOsVersion = [System.Version]::Parse("$($os.Version).$UBR")
    $hostOs = "Unknown/Insider build"
    $bestGenericImageName = Get-BestGenericImageName -onlyMatchingBuilds -filesOnly:$filesOnly
    $isServerHost = $os.ProductType -eq 3

    if ("$baseImage" -eq "") {
        if ("$bestGenericImageName" -eq "") {
            $bestGenericImageName = Get-BestGenericImageName -filesOnly:$filesOnly
            Write-Host "WARNING: Unable to find matching generic image for your host OS. Using $bestGenericImageName"
        }
        $baseImage = $bestGenericImageName
    }
```

could you try to replace this code in New-NavImage.ps1 with this code:

```
    $hostOsVersion = [System.Version]::Parse("$($os.Version).$UBR")
    $hostOs = "Unknown/Insider build"
    $isServerHost = $os.ProductType -eq 3

    if ("$baseImage" -eq "") {
        $bestGenericImageName = Get-BestGenericImageName -onlyMatchingBuilds -filesOnly:$filesOnly
        if ("$bestGenericImageName" -eq "") {
            $bestGenericImageName = Get-BestGenericImageName -filesOnly:$filesOnly
            Write-Host "WARNING: Unable to find matching generic image for your host OS. Using $bestGenericImageName"
        }
        $baseImage = $bestGenericImageName
    }
    else {
        $bestGenericImageName = $baseImage
    }
```

Note the move of $bestGenericImageName = xxx

**Comment by freddydk** [comment]
Removed two other suggestions, I think the best fix for this is to replace this line (line 301 in New-NavImage.ps1)
```
            if ($baseImage -eq $bestGenericImageName) {
```
with this line:
```
            if ($baseImage -like 'mcr.microsoft.com/businesscentral:*') {
```

**Comment by freddydk** [comment]
Please try this and report back, thanks

**Comment by freddydk** [comment]
@ashirokikh - did you try this fix?

**Comment by ashirokikh** [comment]
Thank you @freddydk that worked well! The tag line displays `Generic Tag: 1.0.2.13`

```
...

Pulling mcr.microsoft.com/businesscentral:10.0.19041.1415

Pulling generic image took 4 seconds

   _____                _   _                               _        _
  / ____|              | | (_)                             | |      (_)
 | |     _ __ ___  __ _| |_ _ _ __   __ _    ___ ___  _ __ | |_ __ _ _ _ __   ___ _ __
 | |    | '__/ _ \/ _` | __| | '_ \ / _` |  / __/ _ \| '_ \| __/ _` | | '_ \ / _ \ '__|
 | |____| | |  __/ (_| | |_| | | | | (_| | | (__ (_) | | | | |_ (_| | | | | |  __/ |
  \_____|_|  \___|\__,_|\__|_|_| |_|\__, |  \___\___/|_| |_|\__\__,_|_|_| |_|\___|_|
                                     __/ |
                                    |___/

INFO: Windows 10 21H1/21H2 images are not yet available, using 2004 as these are found to work better than 20H2 on 21H1/21H2
BcContainerHelper is version 4.0.1
BcContainerHelper is not running as administrator
UsePsSession is False
Host is Microsoft Windows 10 Pro - 21H2
Docker Client Version is 20.10.17
Docker Server Version is 20.10.17
Removing Session bctest2
Removing container bctest2
Removing entries from hosts
Removing bctest2 from container hosts file
Removing bctest2-* from container hosts file
Removing Desktop shortcuts
Removing C:\ProgramData\BcContainerHelper\Extensions\bctest2
Fetching all docker images
Fetching all docker volumes
ArtifactUrl and ImageName specified
Image bctest2-image:sandbox-20.0.37253.44689-au-mt already exists
Using image bctest2-image:sandbox-20.0.37253.44689-au-mt
Creating Container bctest2
Style: sandbox
Multitenant: Yes
Version: 20.0.37253.44689
Platform: 20.0.37114.44567
Generic Tag: 1.0.2.13
Container OS Version: 10.0.19041.1415 (2004)
Host OS Version: 10.0.19044.1889 (21H2)
WARNING: Host OS is 21H1 and Container OS is 2004, defaulting to process isolation. If you experience problems, add -isolation hyperv.
Using process isolation
Using locale en-AU
...
```

**Comment by ashirokikh** [comment]
Sorry @freddydk that was after the image was first built without the `useGenericImage` parameter. With the suggested update to the `New-NavImage.ps1` script, it does not build a new image if we add the `useGenericImage` parameter second time. However, if the first image was built with the `useGenericImage` parameter, the second container creation run will still rebuild the image.

First run:
```
...
BcContainerHelper is version 4.0.1
BcContainerHelper is not running as administrator
UsePsSession is False
Host is Microsoft Windows 10 Pro - 21H2
Docker Client Version is 20.10.17
Docker Server Version is 20.10.17
Removing Desktop shortcuts
Fetching all docker images
Fetching all docker volumes
ArtifactUrl and ImageName specified
Building multitenant image bctest2-image:sandbox-20.0.37253.44689-au-mt based on mcr.microsoft.com/businesscentral:10.0.19041.1415-dev with https://bcartifacts.azureedge.net/sandbox/20.0.37253.44689/au
Generic Tag: 1.0.2.4
Container OS Version: 10.0.19041.1415 (2004)
Host OS Version: 10.0.19044.1889 (21H2)
WARNING: Host OS is 21H2 and Container OS is 2004, defaulting to process isolation. If you experience problems, add -isolation hyperv.
Using process isolation
...
```
Second run:
```
...
BcContainerHelper is version 4.0.1
BcContainerHelper is not running as administrator
UsePsSession is False
Host is Microsoft Windows 10 Pro - 21H2
Docker Client Version is 20.10.17
Docker Server Version is 20.10.17
Removing Session bctest2
Removing container bctest2
Removing entries from hosts
Removing bctest2 from container hosts file
Removing bctest2-* from container hosts file
Removing Desktop shortcuts
Removing C:\ProgramData\BcContainerHelper\Extensions\bctest2
Fetching all docker images
Fetching all docker volumes
ArtifactUrl and ImageName specified
Image bctest2-image:sandbox-20.0.37253.44689-au-mt already exists
Image bctest2-image:sandbox-20.0.37253.44689-au-mt has generic Tag 1.0.2.4, should be 1.0.2.13
Building multitenant image bctest2-image:sandbox-20.0.37253.44689-au-mt based on mcr.microsoft.com/businesscentral:10.0.19041.1415-dev with https://bcartifacts.azureedge.net/sandbox/20.0.37253.44689/au
Generic Tag: 1.0.2.4
Container OS Version: 10.0.19041.1415 (2004)
Host OS Version: 10.0.19044.1889 (21H2)
WARNING: Host OS is 21H2 and Container OS is 2004, defaulting to process isolation. If you experience problems, add -isolation hyperv.
Using process isolation
...
```

**Comment by ashirokikh** [comment]
And of course I should have sticked to the original script which used the installed BcContainerHelper module instead of the AL-Go scripts which appear to be using their own instance of the latest version of the module.

Please ignore my previous comments, with the suggested update to the New-NavImage.ps1 script, the outputs look like the below:

First run:
```
...
BcContainerHelper is version 3.0.12
BcContainerHelper is not running as administrator
UsePsSession is False
Host is Microsoft Windows 10 Pro - 21H2
Docker Client Version is 20.10.17
Docker Server Version is 20.10.17
Removing Desktop shortcuts
Fetching all docker images
Fetching all docker volumes
ArtifactUrl and ImageName specified
Building multitenant image bc20autest-image:sandbox-20.0.37253.44689-au-mt based on mcr.microsoft.com/businesscentral:10.0.19041.1415-dev with https://bcartifacts.azureedge.net/sandbox/20.0.37253.44689/au
Pulling latest image mcr.microsoft.com/businesscentral:10.0.19041.1415-dev
10.0.19041.1415-dev: Pulling from businesscentral
Digest: sha256:8f994521bb757c5c09687dd7484c8696b015ee4dde430fb141b1e560162bb6b0
Status: Downloaded newer image for mcr.microsoft.com/businesscentral:10.0.19041.1415-dev
mcr.microsoft.com/businesscentral:10.0.19041.1415-dev
Generic Tag: 1.0.2.13
Container OS Version: 10.0.19041.1415 (2004)
Host OS Version: 10.0.19044.1889 (21H2)
WARNING: Host OS is 21H2 and Container OS is 2004, defaulting to process isolation. If you experience problems, add -isolation hyperv.
Using process isolation
...

```

Second run:
```
...
BcContainerHelper is version 3.0.12
BcContainerHelper is not running as administrator
UsePsSession is False
Host is Microsoft Windows 10 Pro - 21H2
Docker Client Version is 20.10.17
Docker Server Version is 20.10.17
Removing Session bc20auTest
Removing container bc20auTest
Removing entries from hosts
Removing bc20auTest from container hosts file
Removing bc20auTest-* from container hosts file
Removing Desktop shortcuts
Removing C:\ProgramData\BcContainerHelper\Extensions\bc20auTest
Fetching all docker images
Fetching all docker volumes
ArtifactUrl and ImageName specified
Image bc20autest-image:sandbox-20.0.37253.44689-au-mt already exists
Using image bc20autest-image:sandbox-20.0.37253.44689-au-mt
Creating Container bc20auTest
Style: sandbox
Multitenant: Yes
Version: 20.0.37253.44689
Platform: 20.0.37114.44567
Generic Tag: 1.0.2.13
Container OS Version: 10.0.19041.1415 (2004)
Host OS Version: 10.0.19044.1889 (21H2)
WARNING: Host OS is 21H1 and Container OS is 2004, defaulting to process isolation. If you experience problems, add -isolation hyperv.
Using process isolation
...

```
All appears to be working, thanks again.

**Comment by freddydk** [comment]
Thanks - will check this in
Will be part of 4.0.2 prerelease in an hour or so.

**Comment by freddydk** [comment]
Shipped in 4.0.3
